### PR TITLE
Replace BN.cmp with more readable BN.lt/gt

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
@holgerd77 @jwasinger want to review this?

File: lib/runTx.js [hunk 1]
```diff
@@ -92,11 +92,11 @@ module.exports = function (opts, cb) {
     var fromAccount = self.stateManager.cache.get(tx.from)
     var message
 
-    if (!opts.skipBalance && new BN(fromAccount.balance).cmp(tx.getUpfrontCost()) === -1) {
+    if (!opts.skipBalance && new BN(fromAccount.balance).lt(tx.getUpfrontCost())) {
       message = "sender doesn't have enough funds to send tx. The upfront cost is: " + tx.getUpfrontCost().toString() + ' and the sender\'s account only has: ' + new BN(fromAccount.balance).toString()
       cb(new Error(message))
       return
-    } else if (!opts.skipNonce && new BN(fromAccount.nonce).cmp(new BN(tx.nonce)) !== 0) {
+    } else if (!opts.skipNonce && !(new BN(fromAccount.nonce).eq(new BN(tx.nonce)))) {
```
Comment (axic) [inline lib/runTx.js]: Note to self: should introduce `neq` on bn.js

File: lib/opFns.js [hunk 2]
```diff
@@ -137,7 +137,7 @@ module.exports = {
     val = Buffer.from(val) // use clone, don't modify object reference
     var extendOnes = false
 
-    if (k.cmpn(31) <= 0) {
+    if (k.lten(31)) {
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 3]
```diff
@@ -155,26 +155,22 @@ module.exports = {
   // 0x10 range - bit ops
   LT: function (a, b, runState) {
     return Buffer.from([
-      new BN(a)
-        .cmp(new BN(b)) === -1
+      new BN(a).lt(new BN(b))
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 4]
```diff
@@ -155,26 +155,22 @@ module.exports = {
...
     ])
   },
   GT: function (a, b, runState) {
     return Buffer.from([
-      new BN(a)
-        .cmp(new BN(b)) === 1
+      new BN(a).gt(new BN(b))
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 5]
```diff
@@ -155,26 +155,22 @@ module.exports = {
...
     ])
   },
   SLT: function (a, b, runState) {
     return Buffer.from([
-      utils.fromSigned(a)
-        .cmp(utils.fromSigned(b)) === -1
+      utils.fromSigned(a).lt(utils.fromSigned(b))
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 6]
```diff
@@ -155,26 +155,22 @@ module.exports = {
...
     ])
   },
   SGT: function (a, b, runState) {
     return Buffer.from([
-      utils.fromSigned(a)
-        .cmp(utils.fromSigned(b)) === 1
+      utils.fromSigned(a).gt(utils.fromSigned(b))
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 7]
```diff
@@ -761,7 +757,7 @@ function describeLocation (runState) {
 
 function subGas (runState, amount) {
   runState.gasLeft.isub(amount)
-  if (runState.gasLeft.cmpn(0) === -1) {
+  if (runState.gasLeft.ltn(0)) {
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/opFns.js [hunk 8]
```diff
@@ -802,7 +798,7 @@ function subMemUsage (runState, offset, length) {
   // words * 3 + words ^2 / 512
   const cost = words.mul(fee).add(words.mul(words).div(quadCoeff))
 
-  if (cost.cmp(runState.highestMemCost) === 1) {
+  if (cost.gt(runState.highestMemCost)) {
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/runBlock.js [hunk 9]
```diff
@@ -83,7 +83,7 @@ module.exports = function (opts, cb) {
     async.eachSeries(block.transactions, processTx, cb)
 
     function processTx (tx, cb) {
-      var gasLimitIsHigherThanBlock = new BN(block.header.gasLimit).cmp(new BN(tx.gasLimit).add(gasUsed)) === -1
+      var gasLimitIsHigherThanBlock = new BN(block.header.gasLimit).lt(new BN(tx.gasLimit).add(gasUsed))
```
Comment (holgerd77) [inline lib/runBlock.js]: Not sure, what the semantics SHOULD be here, but this also evaluates to ``true`` on equal case, which is not what the variable name ``gasLimitIsHigherThanBlock`` is saying (same as in the old version). Refactoring generally is ok. Would also switch this to a ``gte`` (or ``gt`` ?) comparison for better readability.

File: lib/opFns.js [hunk 10]
```diff
@@ -888,7 +884,7 @@ function makeCall (runState, callOptions, localOpts, cb) {
 
   // check if account has enough ether
   // Note: in the case of delegatecall, the value is persisted and doesn't need to be deducted again
-  if (runState.depth >= fees.stackLimit.v || (callOptions.delegatecall !== true && new BN(runState.contract.balance).cmp(callOptions.value) === -1)) {
+  if (runState.depth >= fees.stackLimit.v || (callOptions.delegatecall !== true && new BN(runState.contract.balance).lt(callOptions.value))) {
```
Comment (holgerd77) [inline lib/opFns.js]: Ok.

File: lib/runCall.js [hunk 11]
```diff
@@ -158,7 +158,7 @@ module.exports = function (opts, cb) {
         var returnFee = results.return.length * fees.createDataGas.v
         var totalGas = results.gasUsed.addn(returnFee)
         // if not enough gas
-        if (totalGas.cmp(gasLimit) <= 0 && results.return.length <= 24576) {
+        if (totalGas.lte(gasLimit) && results.return.length <= 24576) {
```
Comment (holgerd77) [inline lib/runCall.js]: Ok.

File: lib/runCode.js [hunk 12]
```diff
@@ -157,7 +157,7 @@ module.exports = function (opts, cb) {
       var fee = new BN(opInfo.fee)
       // TODO: move to a shared funtion; subGas in opFuns
       runState.gasLeft = runState.gasLeft.sub(fee)
-      if (runState.gasLeft.cmpn(0) === -1) {
+      if (runState.gasLeft.ltn(0)) {
```
Comment (holgerd77) [inline lib/runCode.js]: Ok.

File: lib/runTx.js [hunk 13]
```diff
@@ -28,7 +28,7 @@ module.exports = function (opts, cb) {
     block = new Block()
   }
 
-  if (new BN(block.header.gasLimit).cmp(new BN(tx.gasLimit)) === -1) {
+  if (new BN(block.header.gasLimit).lt(new BN(tx.gasLimit))) {
```
Comment (holgerd77) [inline lib/runTx.js]: Refactoring ok, would be a bit more readable if switched to a ``gte`` comparison (which also has a slightly different semantics than the error message).

File: lib/runTx.js [hunk 14]
```diff
@@ -92,11 +92,11 @@ module.exports = function (opts, cb) {
     var fromAccount = self.stateManager.cache.get(tx.from)
     var message
 
-    if (!opts.skipBalance && new BN(fromAccount.balance).cmp(tx.getUpfrontCost()) === -1) {
+    if (!opts.skipBalance && new BN(fromAccount.balance).lt(tx.getUpfrontCost())) {
```
Comment (holgerd77) [inline lib/runTx.js]: Ok.

File: lib/runTx.js (hunk 1 above)
Comment (holgerd77) [inline lib/runTx.js]: Ok.

File: lib/runTx.js [hunk 15]
```diff
@@ -147,7 +147,7 @@ module.exports = function (opts, cb) {
       // process any gas refund
       var gasRefund = results.vm.gasRefund
       if (gasRefund) {
-        if (gasRefund.cmp(results.gasUsed.divn(2)) === -1) {
+        if (gasRefund.lt(results.gasUsed.divn(2))) {
```
Comment (holgerd77) [inline lib/runTx.js]: Ok.

File: lib/runBlock.js (hunk 9 above)
Comment (axic) [inline lib/runBlock.js]: Why would it evaluate true on equal? That would be `lte`.

Comment (axic) [inline lib/runBlock.js]: It would make sense swapping the arguments for readability but would do that in a separate PR.

Comment (holgerd77) [inline lib/runBlock.js]: Ah sorry, ``true`` to equal if already switched to the equivalent ``gte`` expression:

```
(new BN(tx.gasLimit).add(gasUsed)).gte(new BN(block.header.gasLimit))
```

Which is not the same semantics as ``gasLimitIsHigherThanBlock`` (equivalent would be ``gasLimitIsHigherOrEqualThanBlock``.

Comment (axic) [inline lib/runBlock.js]: Do you want to submit a PR after this is merged to clean up those?

**Comment by axic** [comment]
Thanks @holgerd77 merging it.

File: lib/runBlock.js (hunk 9 above)
Comment (holgerd77) [inline lib/runBlock.js]: Not worth for the switch I suppose. But I think this is a potential consensus issue here.

Just looked it up in the white paper:

bildschirmfoto 2017-08-16 um 12 53 16

So it's really only invalid in ``gt`` case while we have (implicitly) used ``gte`` for the error case. So the ``equal`` case is falsely included there.

Comment (axic) [inline lib/runBlock.js]: Can you address it in a new PR please?

Comment (holgerd77) [inline lib/runBlock.js]: Ok, will do.
